WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainActivity.kt

```diff
@@ -0,0 +1,269 @@
+                            MainViewModel.AppEvent.ShowPermissionRationale -> AlertDialog.Builder(this@MainActivity)
+                                .setMessage(R.string.rationale_for_microphone_permission)
+                                .setPositiveButton(R.string.ok) { _, _ ->
+                                    requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                                }
+                                .setNegativeButton(R.string.cancel) { _, _ -> }
+                                .create()
+                                .show()
+                            MainViewModel.AppEvent.ShowSpeakerNotSupported -> AlertDialog.Builder(this@MainActivity)
+                                .setMessage(R.string.no_speaker_supported)
+                                .setPositiveButton(R.string.ok) { _, _ -> }
+                                .create()
+                                .show()
+                        }
+                    }
+                    .collect()
+            }
+        }
+    }
+
+    /**
+     * The primary update method based on the [AppState].
+     *
+     * This method suspends, as we do suspending work based on the [appState].
+     */
+    private suspend fun onAppState(appState: AppState) {
+        Log.d(TAG, "New app state is: $appState")
+
+        // Only reset the progress bar if we are transitioning to recording. This allows for a nicer
+        // transition away from recording
+        if (appState == AppState.Recording) {
+            binding.progressBar.progress = 0
+        }
+
+        @IdRes val motionLayoutState: Int
+        val transitionInstantly: Boolean
+
+        when (appState) {
+            is AppState.Ready -> {
+                motionLayoutState = R.id.allMinimized
+                transitionInstantly = appState.transitionInstantly
+            }
+            AppState.PlayingVoice -> {
+                motionLayoutState = R.id.playExpanded
+                transitionInstantly = false
+            }
+            AppState.PlayingMusic -> {
+                motionLayoutState = R.id.musicExpanded
+                transitionInstantly = false
+            }
+            AppState.Recording -> {
+                motionLayoutState = R.id.micExpanded
+                transitionInstantly = false
+            }
+        }
+
+        binding.outerCircle.transitionToState(motionLayoutState, transitionInstantly)
+        binding.outerCircle.awaitState(motionLayoutState)
+
+        when (appState) {
+            is AppState.Ready -> {
+                binding.mic.contentDescription = getString(R.string.record)
+                binding.play.contentDescription = getString(R.string.play_recording)
+                binding.music.contentDescription = getString(R.string.play_music)
+            }
+            AppState.PlayingVoice -> {
+                binding.play.contentDescription = getString(R.string.stop_playing_recording)
+
+                soundRecorder!!.play()
```

Comment:
Does `soundRecorder` need to be nullable? It doesn't seem like it holds any resources open when it's not playing or recording.

Reply:
+1 -- Even if it _did_ need to hold resources while it was open, it seems like it would be nicer to still have it be non-null and have `SoundRecorder` observe the lifecycle of the activity to take and release those resources internally, rather than having that logic in the activity itself.

Reply:
Makes sense to me! Converted to a `lateinit var` in 2fb0329